natpmp/client.go

```diff
@@ -0,0 +1,204 @@
+package natpmp
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+)
+
+// Version is the expected protocol version for NAT-PMP.
+const Version = 0
+
+// ErrProtocol indicates that a NAT gateway returned a response that violates
+// the NAT-PMP protocol.
+var ErrProtocol = errors.New("natpmp: protocol error")
+
+//go:generate stringer -type=Error -output=strings.go
+
+// An Error is a NAT-PMP protocol result code which indicates that an operation
+// has failed.
+type Error int
+
+// Possible Errors as defined in RFC 6886, section 3.5.
+const (
+	UnsupportedVersion Error = 1
+	NotAuthorized      Error = 2
+	NetworkFailure     Error = 3
+	OutOfResources     Error = 4
+	UnsupportedOpcode  Error = 5
+
+	// Although success is technically a result code, we don't expose it
+	// directly to the user because a successful operation returns nil error.
+	success Error = 0
+)
+
+// Error implements error.
+func (e Error) Error() string {
+	return fmt.Sprintf("natpmp: result %d: %s", e, e.String())
+}
+
+// A Client is a NAT-PMP client which can communicate with a NAT gateway.
```

Comment:
Something to consider for future changes: NAT-PMP and PCP are so closely related that it may be worth thinking of them as two versions of the same protocol. In which case maybe there should be a Client that's just "do port mapping things", and it selects the appropriate protocol version under the hood autonomously?

Not a strong opinion, just pondering the shape of the API.

Reply:
Yep agreed, see https://github.com/inetaf/nat/pull/1#discussion_r462522434 for my tentative plans.